elki-core-math/src/main/java/elki/math/statistics/tests/mcde/MWPTest.java

```diff
@@ -0,0 +1,191 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package elki.math.statistics.tests.mcde;
+
+import elki.utilities.datastructures.arraylike.NumberArrayAdapter;
+import elki.utilities.documentation.Reference;
+import elki.utilities.exceptions.AbortException;
+import elki.utilities.random.RandomFactory;
+
+import java.lang.reflect.Array;
+
+import static elki.math.statistics.distribution.NormalDistribution.erf;
+
+/**
+ * Implementation of Mann-Whitney U test returning the p-value (not the test
+ * statistic, thus MWP) for MCDEDependenceMeasure. Implements algorithm 1 and 3
+ * of reference paper.
+ *
+ * @author Alan Mazankiewicz
+ * @author Edouard Fouché
+ */
+
+@Reference(authors = "Edouard Fouché, Klemens Böhm", //
+    title = "Monte Carlo Density Estimation", //
+    booktitle = "Proc. 2019 ACM Int. Conf. on Scientific and Statistical Database Management (SSDBM 2019)", url = "https://doi.org/10.1145/3335783.3335795", //
+    bibkey = "DBLP:conf/ssdbm/FoucheB19")
+
+public class MWPTest extends MCDETest<MWPTest.MwpIndex> {
+
+  private static final long two_times_sqrt_max_long = 2 * (long) Math.sqrt(Long.MAX_VALUE);
+
+  /**
+   * Share of instances in marginal restriction (reference dimension).
+   * Note that in the original paper alpha = beta and as such there is no explicit distinction between the parameters.
+   */
+  protected double beta;
+
+  /**
+   * Random generator.
+   */
+  protected RandomFactory rnd;
+
+  /**
+   * Structure to hold values needed for computing MWP in MCDEDependeneMeasure.
+   * Values get computed during index creation.
+   */
+
+  public class MwpIndex extends MCDETest.RankStruct {
+    long correction;
+
+    public double adjusted;
+
+    public MwpIndex(int index, double adjusted, long correction) {
+      super(index);
+      this.correction = correction;
+      this.adjusted = adjusted;
+    }
+  }
+
+  /**
+   * Static Constructor - use this one
+   */
+
+  public static final MWPTest STATIC = new MWPTest();
+
+  /**
+   * Private Constructor - use {@link #STATIC}
+   */
+
+  protected MWPTest(){
+  }
+
+  /**
+   * Computes Corrected Rank Index as described in Algorithm 1 of reference paper.
+   * The notation follows ELKI convention if applicable, else we use the notation from the reference paper.
+   *
+   * @param adapter ELKI NumberArrayAdapter Subclass
+   * @param data    One dimensional array containing one dimension of the data
+   * @param idx     Return value of sortedIndex()
+   * @return Array of doubles, 3 subsequent values being assigned to one data instance.
+   * Containing sorted (ascending) row numbers, adjusted ranks and tying value corrections
+   * as required by MWP test. Example:
+   * double[] corrected_ranks = corrected_ranks(...);
```

Comment:
Code examples should be wrapped in `{@code }` so they do not break when auto-formatted. The `@return` tag should have a short explanation, the usage example should go to the main JavaDoc of the method instead.